Error

```
java.lang.RuntimeException: Could not set max YAML line width
        at pl.mikigal.config.BukkitConfiguration.overrideMaxLineWidth(BukkitConfiguration.java:222) ~[BetterKingdoms.jar:?]
        at pl.mikigal.config.BukkitConfiguration.saveToString(BukkitConfiguration.java:113) ~[BetterKingdoms.jar:?]
        at org.bukkit.configuration.file.FileConfiguration.save(FileConfiguration.java:65) ~[?:git-UniverseSpigot-"252422e"]
        at pl.mikigal.config.BukkitConfiguration.save(BukkitConfiguration.java:171) ~[BetterKingdoms.jar:?]
        at pl.mikigal.config.ConfigInvocationHandler.updateConfigFile(ConfigInvocationHandler.java:254) ~[BetterKingdoms.jar:?]
        at pl.mikigal.config.ConfigInvocationHandler.<init>(ConfigInvocationHandler.java:41) ~[BetterKingdoms.jar:?]
        at pl.mikigal.config.ConfigAPI.init(ConfigAPI.java:71) ~[BetterKingdoms.jar:?]
        at pl.mikigal.config.ConfigAPI.init(ConfigAPI.java:91) ~[BetterKingdoms.jar:?]
        at dev.isnow.betterkingdoms.config.ConfigManager.init(ConfigManager.java:27) ~[BetterKingdoms.jar:?]
        at dev.isnow.betterkingdoms.config.ConfigManager.<init>(ConfigManager.java:22) ~[BetterKingdoms.jar:?]
        at dev.isnow.betterkingdoms.BetterKingdoms.onEnable(BetterKingdoms.java:44) ~[BetterKingdoms.jar:?]
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:287) ~[?:git-UniverseSpigot-"252422e"]
        at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.enablePlugin(PaperPluginInstanceManager.java:188) ~[?:git-UniverseSpigot-"252422e"]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.enablePlugin(PaperPluginManagerImpl.java:104) ~[?:git-UniverseSpigot-"252422e"]
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:507) ~[?:git-UniverseSpigot-"252422e"]
        at org.bukkit.craftbukkit.v1_20_R3.CraftServer.enablePlugin(CraftServer.java:642) ~[?:git-UniverseSpigot-"252422e"]
        at org.bukkit.craftbukkit.v1_20_R3.CraftServer.enablePlugins(CraftServer.java:553) ~[?:git-UniverseSpigot-"252422e"]
        at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:678) ~[?:git-UniverseSpigot-"252422e"]
        at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:438) ~[?:git-UniverseSpigot-"252422e"]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:323) ~[?:git-UniverseSpigot-"252422e"]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1139) ~[?:git-UniverseSpigot-"252422e"]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:324) ~[?:git-UniverseSpigot-"252422e"]
        at java.lang.Thread.run(Thread.java:840) ~[?:?]
Caused by: java.lang.NoSuchFieldException: yamlOptions
        at java.lang.Class.getDeclaredField(Class.java:2610) ~[?:?]
        at pl.mikigal.config.BukkitConfiguration.overrideMaxLineWidth(BukkitConfiguration.java:217) ~[BetterKingdoms.jar:?]
        ... 22 more
```

```java
@Getter
public class ConfigManager {
    private final DatabaseConfig databaseConfig;

    public ConfigManager(BetterKingdoms plugin) {
        // Another api issue...
        ConfigAPI.registerSerializer(DatabaseType.class, new DatabaseTypeSerializer());

        databaseConfig = (DatabaseConfig) init(DatabaseConfig.class, false, plugin);
    }

    private Config init(final Class<? extends Config> clazz, final boolean color, final BetterKingdoms plugin) {
        return ConfigAPI.init(
                (Class<? extends Config>) clazz,
                NameStyle.CAMEL_CASE,
                CommentStyle.INLINE,
                color,
                plugin
        );
    }
}
```

```java
public enum DatabaseType {
    MYSQL,
    MARIADB,
    H2,
    POSTGRES,
    SQLite,
    Oracle;
}
```

```java
@ConfigName("database.yml")
public interface DatabaseConfig extends Config {
    void setDatabase(Database database);

    default Database getDatabase() {
        return new Database("127.0.0.1", "sa", "", "betterkingdoms", DatabaseType.H2);
    }
}
```

```java
@Getter
@Setter
public class Database implements Serializable {
    private String ip;
    private String username;
    private String password;
    private String databaseName;
    private DatabaseType databaseType;

    public Database() {} // Config system stuff

    public Database(final String ip, final String username, final String password, final String databaseName, final DatabaseType databaseType) {
        this.ip = ip;
        this.username = username;
        this.password = password;
        this.databaseName = databaseName;
        this.databaseType = databaseType;
    }
}
```

Also another issue, I have to write a serlizers for enums because apparently it can't find a default constructor for them
```java
public class DatabaseTypeSerializer extends Serializer<DatabaseType> {
    @Override
    protected void saveObject(final String s, final DatabaseType databaseType, final BukkitConfiguration bukkitConfiguration) {
        bukkitConfiguration.set(s, databaseType.toString().toUpperCase(Locale.ROOT));
    }

    @Override
    public DatabaseType deserialize(final String s, final BukkitConfiguration bukkitConfiguration) {
        return DatabaseType.valueOf(bukkitConfiguration.getString(s).toUpperCase(Locale.ROOT));
    }
}
```

--- a/src/main/java/pl/mikigal/config/BukkitConfiguration.java
+++ b/src/main/java/pl/mikigal/config/BukkitConfiguration.java
@@ -1,5 +1,6 @@
 package pl.mikigal.config;
 
+import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
 import org.yaml.snakeyaml.DumperOptions;
@@ -11,6 +12,7 @@ import pl.mikigal.config.serializer.Serializers;
 import pl.mikigal.config.style.CommentStyle;
 import pl.mikigal.config.style.NameStyle;
 import pl.mikigal.config.util.ConversionUtils;
+import pl.mikigal.config.util.ReflectionUtils;
 import pl.mikigal.config.util.TypeUtils;
 
 import java.io.*;
@@ -40,6 +42,12 @@ public class BukkitConfiguration extends YamlConfiguration {
 	 */
 	private final Map<String, Object> cache;
 	private final Map<String, String> comments;
+
+	/**
+	 *
+	 * Reflection Path
+	 */
+	private final String dumperFieldName = ReflectionUtils.isVeryNewVersion() ? "yamlDumperOptions" : "yamlOptions";
 
 	public BukkitConfiguration(File directory, File file, NameStyle nameStyle, CommentStyle commentStyle,
 							   boolean automaticColorStrings, String configComment) {
@@ -214,7 +222,7 @@ public class BukkitConfiguration extends YamlConfiguration {
 	 */
 	private void overrideMaxLineWidth() {
 		try {
-			Field yamlOptionsField = YamlConfiguration.class.getDeclaredField("yamlOptions");
+			Field yamlOptionsField = YamlConfiguration.class.getDeclaredField(dumperFieldName);
 			yamlOptionsField.setAccessible(true);
 			DumperOptions yamlOptions = (DumperOptions) yamlOptionsField.get(this);
 			yamlOptions.setWidth(Integer.MAX_VALUE);

--- a/src/main/java/pl/mikigal/config/util/ReflectionUtils.java
+++ b/src/main/java/pl/mikigal/config/util/ReflectionUtils.java
@@ -105,4 +105,13 @@ public class ReflectionUtils {
 		int version = Integer.parseInt(Bukkit.getBukkitVersion().split("\\.")[1].split("-")[0]);
 		return version >= 12;
 	}
+
+	/**
+	 * Check bukkit versions
+	 * @return false for Minecraft 1.18 or older, true for 1.18 or newer
+	 */
+	public static boolean isVeryNewVersion() {
+		int version = Integer.parseInt(Bukkit.getBukkitVersion().split("\\.")[1].split("-")[0]);
+		return version >= 18;
+	}
 }

